### PR TITLE
boards: arm: arduino_opta_m4: adds initial support to Arduino OPTA PLC on the M4 core

### PR DESCRIPTION
**Pull request by uLipe**
The idea is to coexist both Arduino Core, responsible to do the PLC runtime, and use Zephyr on M4 as a coprocessor core responsible to real-time tasks and offload the M7.

### CODE REVIEW

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (facchinm) [inline boards/arm/arduino_opta_m4/doc/index.rst:21]: The feature list looks like a copy paste mistake from the EVK :slightly_smiling_face:

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (facchinm) [inline boards/arm/arduino_opta_m4/doc/index.rst:137]: The PLC binary running on the M7 could grow a bit in the future, maybe it's better to dedicate a bit less flash to the M4? I think 512K would be sensible. Then on the Arduino side we can just call bootM4() https://github.com/arduino/ArduinoCore-mbed/blob/main/variants/OPTA/variant.cpp#L380-L384  with the right address

**Review by facchinm — CHANGES_REQUESTED** [review]

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (uLipe) [reply boards/arm/arduino_opta_m4/doc/index.rst:137]: Yes, I noticed in Arduino IDE it is possible to do flash partitioning in order to leave 512k of flash memory to the M4 instead of equal split.

Comment (uLipe) [reply boards/arm/arduino_opta_m4/doc/index.rst:130]: Moved the flash1 reagion 512K ahead, and left 1.5MB for Arduino and 512K upper for the M4 + Zephyr.

**Comment by uLipe** [comment]
@nashif the failure is reporting the miss of license identifier, but the file has it, do you know any hint for this failure?

**Comment by kartben** [comment]
> @nashif the failure is reporting the miss of license identifier, but the file has it, do you know any hint for this failure?

It looks incomplete to me, as it's missing copyright info.

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (facchinm) [reply boards/arm/arduino_opta_m4/doc/index.rst:130]: Super, I've seen the change in the dts but the documentation still uses 0x81000000 as default address

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (facchinm) [inline boards/arm/arduino_opta_m4/doc/index.rst:24]: Still some HW configuration to be tuned:
* USB FS
* 128Mb QUAD-SPI flash

All other info is correct :slightly_smiling_face:

**Review by facchinm — CHANGES_REQUESTED** [review]

**Comment by uLipe** [comment]
> > @nashif the failure is reporting the miss of license identifier, but the file has it, do you know any hint for this failure?
> 
> It looks incomplete to me, as it's missing copyright info.

Ahh thanks @kartben I noticed it now.

**Comment by uLipe** [comment]
It seems the test failure is not related to this PR.

File: boards/arm/arduino_opta_m4/arduino_opta_m4.dts
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/arduino_opta_m4.dts:16]: nit: these are normally all lowercase, `arduino,opta`

File: boards/arm/arduino_opta_m4/arduino_opta_m4.dts
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/arduino_opta_m4.dts:27]: Not a great way of doing it, leave back `flash1` as it is and define a `fixed-partitions` with an entry only using the second half, then set it as `zephyr,code-partition` and enable `CONFIG_USE_DT_CODE_PARTITION=y` in the board configuration. That's how it's normally handled for external bootloaders and this type of situation.

File: boards/arm/arduino_opta_m4/arduino_opta_m4.dts
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/arduino_opta_m4.dts:36]: `status_led_1: led_1` (add a space, same for the other below)

File: boards/arm/arduino_opta_m4/arduino_opta_m4.dts
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/arduino_opta_m4.dts:38]: these are not really used for anything anymore AFAIK, unless you want the labeles set for whatever reasons I suggest dropping them, makes it easy to copypaste the block around without stale data

File: boards/arm/arduino_opta_m4/arduino_opta_m4.dts
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/arduino_opta_m4.dts:39]: not needed, there's nothing defining this node as `disabled`, drop

File: boards/arm/arduino_opta_m4/arduino_opta_m4.dts
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/arduino_opta_m4.dts:64]: same as above, drop

File: boards/arm/arduino_opta_m4/arduino_opta_m4.dts
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/arduino_opta_m4.dts:82]: is that correct? other M4s have this running at 240

File: boards/arm/arduino_opta_m4/arduino_opta_m4_defconfig
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/arduino_opta_m4_defconfig:7]: stale comment, drop

File: boards/arm/arduino_opta_m4/arduino_opta_m4_defconfig
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/arduino_opta_m4_defconfig:24]: no point having these ready if there's no `zephyr,console` defined, drop

File: boards/arm/arduino_opta_m4/board.cmake
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/board.cmake:5]: does this actually work? how do you load the firmware on the board?

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/doc/index.rst:52]: drop, not defined in the dts file

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/doc/index.rst:64]: prob should be dropped as well?

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (fabiobaltieri) [inline boards/arm/arduino_opta_m4/doc/index.rst:87]: nit, I'd use `Status LED` and `User button`

File: boards/arm/arduino_opta_m4/board.cmake
Comment (uLipe) [reply boards/arm/arduino_opta_m4/board.cmake:5]: Yes all the stm32h7 has a ROM DFU bootloader ( also the Arduino opta bootloader is DFU ), to load the firmware just connect the board to the host computer, hit the reset button shortly two times to make the Opta in DFU mode, then issue west flash.

This procedure is covered into the documentation, it is the same to other stm32h7 arduino boards :)

Comment (fabiobaltieri) [reply boards/arm/arduino_opta_m4/board.cmake:5]: Thanks, just checking, I'm unfamiliar with the platform. :-)

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (uLipe) [reply boards/arm/arduino_opta_m4/doc/index.rst:64]: This is something I find ambiguous sometimes, because although both SPI and UART are not in the DTS file, saying they are not supported is not entirely true, since just an overlay fragment makes it to be usable, so put on the boarrd document that other features not mentioned are not supported may discourage users to try to use other features that are actually well supported by the SoC that is not mentioned in its board documentation, so I'm wondering if  we can put on this table two extra columns called supported by SoC, supported ready on the Board.

Anyway, it is just a personal opinion, for now I will remove both UART and SPI :)

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by uLipe** [comment]
@facchinm , please have a look on latest changes :)

**Comment by uLipe** [comment]
Any updates here folks? Thank you :)

**Comment by fabiobaltieri** [comment]
@facchinm @pillo79 could you take another look?

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (facchinm) [inline boards/arm/arduino_opta_m4/doc/index.rst:24]: Replace with `16MB QSPI Flash`

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (facchinm) [inline boards/arm/arduino_opta_m4/doc/index.rst:27]: 8 x analog inputs

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (facchinm) [inline boards/arm/arduino_opta_m4/doc/index.rst:26]: Reset button is separate and accessible through a pinhole

**Review by facchinm — CHANGES_REQUESTED** [review]
Code is good, just some changes in the documentation

**Comment by uLipe** [comment]
Documentation changes addressed thanks @facchinm

File: boards/arm/arduino_opta_m4/arduino_opta_m4.dts
Comment (kartben) [inline boards/arm/arduino_opta_m4/arduino_opta_m4.dts:14]:
```suggestion
	compatible = "arduino,opta-m4";
```

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (kartben) [inline boards/arm/arduino_opta_m4/doc/index.rst:18]:
```suggestion
For Zephyr RTOS, only the M4 is supported for now, making the M7 run the PLC
```

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (kartben) [inline boards/arm/arduino_opta_m4/doc/index.rst:77]:
```suggestion
Default Zephyr Peripheral Mapping
---------------------------------
```

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (kartben) [inline boards/arm/arduino_opta_m4/doc/index.rst:90]:
```suggestion
is driven at 240MHz. PLL clock is fed by a 25MHz high speed external clock.
```

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (kartben) [inline boards/arm/arduino_opta_m4/doc/index.rst:110]:
```suggestion
Applications for the ``arduino_opta_m4`` use the regular Zephyr build commands
```

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (kartben) [inline boards/arm/arduino_opta_m4/doc/index.rst:117]:
```suggestion
option bytes configuration. The OPTA has a DFU capable bootloader which
```

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (kartben) [inline boards/arm/arduino_opta_m4/doc/index.rst:120]:
```suggestion
can be accessed by connecting the device to the USB, and then pressing
the RESET button shortly twice; the RESET-LED on the board will fade
indicating the board is in bootloader mode.
```

**Review by kartben — CHANGES_REQUESTED** [review]
a few more mostly doc-related remarks.

Also, you may want to set your editor to wrap .rst file content at 100 characters per line, right now the file seems to have inconsistent wrapping.

**Comment by uLipe** [comment]
Any hints for the build failure in the compliance checks?

**Comment by kartben** [comment]
> Any hints for the build failure in the compliance checks?

@fabiobaltieri is on it. A regression introduced with #58149

**Comment by fabiobaltieri** [comment]
> @fabiobaltieri is on it. A regression introduced with #58149

Fixed, thanks for the heads up.

**Comment by uLipe** [comment]
Hello folks, @fabiobaltieri , @facchinm , do you need any other adjust on this PR?

**Review by fabiobaltieri — DISMISSED** [review]
Sorry lost this in the noise, @facchinm, @kartben can you take another look? You both have a blocker here

**Comment by uLipe** [comment]
@facchinm , @kartben, could you please have another look? I'm just waiting your go signal.

Thank you.

File: boards/arm/arduino_opta_m4/doc/img/arduino_opta.jpeg
Comment (kartben) [inline boards/arm/arduino_opta_m4/doc/img/arduino_opta.jpeg:1]: nit: Could you use this image instead? https://drive.google.com/file/d/1NQO53cxg47KEIaAWMMwYnpYzD-McWyF4/view?usp=drive_link
This one is pretty low quality :/

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (kartben) [inline boards/arm/arduino_opta_m4/doc/index.rst:9]:
```suggestion
The Arduino Opta® is a secure micro Programmable Logic Controller (PLC)
```

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (kartben) [inline boards/arm/arduino_opta_m4/doc/index.rst:110]:
```suggestion
Applications for the ``arduino_opta_m4`` use the regular Zephyr build commands.
```

**Review by kartben — DISMISSED** [review]
+1 as I feel bad asking you to make changes now, but if you have a chance to update the image that would certainly be appreciated :)

**Comment by pillo79** [comment]
Thanks for this! I tested it here and confirm it works fine. :+1:
Just made a couple fixes:
* the DTS did not specify the `zephyr,code-partition` alias so it kept linking at `0x08100000` (start of `flash1`)
* added a tiny description to mention the code that needs to be running on the M7 side for the M4 to be loaded and started properly.

Feel free to merge this in your commit for inclusion!

File: boards/arm/arduino_opta_m4/doc/index.rst
Comment (kartben) [inline boards/arm/arduino_opta_m4/doc/index.rst:159]:
```suggestion
 .. code-block:: cpp

    #include <RPC.h>

    void setup() {
        RPC.begin();
    }

    void loop() { }
```

**Review by kartben — CHANGES_REQUESTED** [review]

**Comment by pillo79** [comment]
Latest iteration merged in @kartben's comments (not sure how I managed to dismiss them via a git operation... :man_shrugging:). Also couldn't access your image link, but replaced with a stock one.

LGTM, @uLipe if my changes are OK for you please squash them so we can add this support. Thanks again!

**Comment by kartben** [comment]
> Latest iteration merged in @kartben's comments (not sure how I managed to dismiss them via a git operation... 🤷‍♂️). Also couldn't access your image link, but replaced with a stock one.

oops, my bad. I've fixed the permissions. Your stock image works for sure. The one I had is essentially the same view but in webp format and a transparent background, which is always better for people using the dark them, but I'm fine either way :) Thanks for the fixups!

**Comment by uLipe** [comment]
@pillo79 @kartben , just squashed the fixups, just waiting for the CI, also I did some local tests and still working like a charm.

Please do a new round to check if we can get this merged.

Thank you!

**Review by fabiobaltieri — APPROVED** [review]

**Review by pillo79 — APPROVED** [review]
Tested on HW here! :+1:

**Review by kartben — APPROVED** [review]

**Comment by uLipe** [comment]
@sylvioalves it might be interesting for you :)

**Comment by fabiobaltieri** [comment]
@facchinm ping

**Review by facchinm — APPROVED** [review]

**Pull request merged**
